include/graaflib/algorithm/graph_isomorphism.h

```diff
@@ -0,0 +1,482 @@
+#pragma once
+
+#include <algorithm>
+#include <graaflib/types.h>
+#include <graaflib/graph.h>
+#include <vector>
+#include <unordered_map>
+#include <optional>
+#include <unordered_set>
+#include <array>
+
+namespace graaf::algorithm {
+
+// Type aliases
+using Mapping = std::unordered_map<vertex_id_t, vertex_id_t>;
+using Candidate_pairs = std::vector<std::pair<vertex_id_t, vertex_id_t>>;
+using Core_set = std::vector<int>;
+using Terminal_set = std::vector<int>;
+
+/**
+ * @brief Checks if two graphs are isomorphic.
+ *
+ * This function checks whether two given graphs are isomorphic, meaning
+ * they are structurally identical up to vertex and edge relabeling.
+ * If the graphs are isomorphic, the function returns an optional containing
+ * a mapping from the vertices of the first graph to the vertices of the second graph.
+ * If the graphs are not isomorphic, the function returns std::nullopt.
+ *
+ * @tparam V The vertex type of the graphs.
+ * @tparam E The edge type of the graphs.
+ * @tparam T The graph type of the graphs (directed or undirected).
+ * @param lhs The first graph to compare.
+ * @param rhs The second graph to compare.
+ * @return An optional containing a mapping from the vertices of lhs to those of rhs if isomorphic; otherwise std::nullopt.
+ */
+template <typename V, typename E, graph_type T>
+std::optional<Mapping>
+check_isomorphism(
```

Comment:
This is the only function users are supposed to call, correct? All the other definitions should be hidden in the `.tpp` file then.

Reply:
Do the other functions that will not be called by the user need comments explaining how they work?